src/ray/util/compat.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/compat.h"
+
+#include <cstring>
+
+#include "ray/util/logging.h"
+
+#if defined(__APPLE__) || defined(__linux__)
+#include <unistd.h>
+#elif defined(_WIN32)
+#include <windows.h>
+#endif
+
+namespace ray {
+
+#if defined(__APPLE__) || defined(__linux__)
+Status CompleteWrite(MEMFD_TYPE_NON_UNIQUE fd, const char *data, size_t len) {
+  const ssize_t ret = write(fd, data, len);
+  if (ret == -1) {
+    return Status::IOError("") << "Fails to write to file because " << strerror(errno);
+  }
+  if (ret != static_cast<ssize_t>(len)) {
+    return Status::IOError("") << "Fails to write all requested bytes, requests to write "
+                               << len << " bytes, but actually write " << ret << " bytes";
+  }
+  return Status::OK();
+}
+void Flush(MEMFD_TYPE_NON_UNIQUE fd) {
```

Comment:
Return Status like others.

Reply:
No fsyncgate is not recoverable and should exit abnormally directly

Reply:
I think as a util, caller should decide whether it wants to exit abnormally or do something else. Otherwise why fsyncgate doesn't exit abnormally inside itself.

Reply:
> Otherwise why fsyncgate doesn't exit abnormally inside itself

What do you mean? It's a behavior, not a function call

Reply:
Are you saying why `fsync` doesn't exit itself? It's because `fsync` behavior differs varies among filesystems, for example, ext-4 is known to clear dirty bit after a failed sync behavior, not sure about other fs.

Interface-wise, POSIX doesn't care about fs type.

Reply:
My reasoning for exit directly, if a sync fails, the filesystem is already in a bad state for current process, and shouldn't be used later on.

Reply:
> I think as a util, caller should decide whether it wants to exit abnormally or do something else

... only if there're options to handle error (i.e. log and continue, propagate, exit)

Reply:
Per https://linux.die.net/man/2/fdatasync, it can return EBADF (fd is not a valid open file descriptor) and this is something caller can handle.

Reply:
I believe postgres panics for all errors: https://github.com/postgres/postgres/blob/43a15eb9400dba2b0b97be72d1a3745a6a6f7136/src/backend/access/transam/xlog.c#L8734-L8745

Reply:
The only err code gets retries is signal interrupt:
https://github.com/postgres/postgres/blob/43a15eb9400dba2b0b97be72d1a3745a6a6f7136/src/backend/storage/file/fd.c#L475-L493

Reply:
If we really need to handle it via mimic the postgres and tigerbeetles' behavior, we will differentiate (1) windows vs unix; (2) error codes, I would like to keep the code simple, so just panic here.

Reply:
I think they assumed fd is always valid. But we are writing a utility, I don't think we should assume that.

Reply:
We can also declare the util function is used inside of ray (the filesystem function is not exposed externally);
anyway I updated the return value, PTAL